mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala

```diff
@@ -208,3 +225,224 @@ class EMLDAOptimizer extends LDAOptimizer{
+      docs.foreach(doc =>{
+        val termCounts = doc._2
+        val (ids, cts) = termCounts match {
+          case v: DenseVector => (((0 until v.size).toList), v.values)
+          case v: SparseVector => (v.indices.toList, v.values)
+          case v => throw new IllegalArgumentException("Do not support vector type " + v.getClass)
+        }
+
+        // Initialize the variational distribution q(theta|gamma) for the mini-batch
+        var gammad = new Gamma(100, 1.0 / 100.0).samplesVector(k).t // 1 * K
+        var Elogthetad = digamma(gammad) - digamma(sum(gammad))     // 1 * K
+        var expElogthetad = exp(Elogthetad)                         // 1 * K
+        val expElogbetad = expElogbeta(::, ids).toDenseMatrix       // K * ids
+
+        var phinorm = expElogthetad * expElogbetad + 1e-100         // 1 * ids
+        var meanchange = 1D
+        val ctsVector = new BDV[Double](cts).t                      // 1 * ids
+
+        // Iterate between gamma and phi until convergence
+        while (meanchange > 1e-5) {
+          val lastgamma = gammad
+          //        1*K                  1 * ids               ids * k
+          gammad = (expElogthetad :* ((ctsVector / phinorm) * (expElogbetad.t))) + alpha
+          Elogthetad = digamma(gammad) - digamma(sum(gammad))
+          expElogthetad = exp(Elogthetad)
+          phinorm = expElogthetad * expElogbetad + 1e-100
+          meanchange = sum(abs(gammad - lastgamma)) / k
+        }
+
+        val m1 = expElogthetad.t.toDenseMatrix.t
+        val m2 = (ctsVector / phinorm).t.toDenseMatrix
+        val outerResult = kron(m1, m2) // K * ids
+        for (i <- 0 until ids.size) {
+          stat(::, ids(i)) := (stat(::, ids(i)) + outerResult(::, i))
+        }
+        stat
+      })
+      Iterator(stat)
+    })
+
+    val batchResult = stats.reduce(_ += _)
+    update(batchResult, iteration, batchSize)
+    batch.unpersist()
+    this
+  }
+
+  private[clustering] override def getLDAModel(iterationTimes: Array[Double]): LDAModel = {
+    new LocalLDAModel(Matrices.fromBreeze(lambda).transpose)
+  }
+
+  private def update(raw: BDM[Double], iter:Int, batchSize: Int): Unit ={
+
+    val tau_0 = this.getTau_0
+    val kappa = this.getKappa
+
+    // weight of the mini-batch.
+    val weight = math.pow(tau_0 + iter, -kappa)
+
+    // This step finishes computing the sufficient statistics for the M step
+    val stat = raw :* expElogbeta
+
+    // Update lambda based on documents.
+    lambda = lambda * (1 - weight) + (stat * D.toDouble / batchSize.toDouble + eta) * weight
+    Elogbeta = dirichlet_expectation(lambda)
+    expElogbeta = exp(Elogbeta)
+  }
+
+  private def getGammaMatrix(row:Int, col:Int): BDM[Double] ={
+    val gammaRandomGenerator = new Gamma(100, 1.0 / 100.0)
+    val temp = gammaRandomGenerator.sample(row * col).toArray
+    (new BDM[Double](col, row, temp)).t
+  }
+
+  private def dirichlet_expectation(alpha : BDM[Double]): BDM[Double] = {
```

Comment:
Please add a little doc, even though it's an internal method

Naming: please use camelCase instead of underscore